In [37]:
from pymongo import MongoClient
client = MongoClient()
db = client.whosampled

import pandas as pd

df = pd.DataFrame(list(db.main_redo.find()))

import itertools

In [38]:
#Turn sampled_song_producer from list
df['sampled_song_producer'] = df.sampled_song_producer.apply( lambda x: ''.join(x))

df.drop(columns= "_id", inplace=True)

#Only 300 or so dupes
df.drop_duplicates(['new_song_name', 'new_song_producer', 'sampled_artist'], inplace=True)

In [39]:
df.shape

(15809, 16)

There are 15809 unique records. There are 1324 producers. There are 7789 songs sampled.

In [40]:
sampled_songs = list(df.drop_duplicates('sampled_song_name')['sampled_song_name'])
new_song_producers = list(df.drop_duplicates('new_song_producer')['new_song_producer'])

In [56]:
# Make new df with zeros everywhere using the columns and rows above.
utility_matrix = pd.DataFrame(0, index=sampled_songs, columns= new_song_producers).to_dict()


In [ ]:
def make_util():
    for producer, songs in utility_matrix.items():
        for song in songs:
            #return producer, song
            cond1 = df.sampled_song_name == song
            cond2 = df.new_song_producer == producer
            if len(df[cond1 & cond2]) > 0:
                utility_matrix[producer][song] = 1
make_util()
    

In [31]:
utility_matrix.head()

It's a New Day  We Want to Parrty, Parrty, Parrty  \
The 45 King                0                                  0   
Marley Marl                0                                  0   
4th Disciple               0                                  0   
RZA                        0                                  0   
Method Man                 0                                  0   

              Get Up, Get Into It, Get Involved  Be Black Baby  Grab Bag  \
The 45 King                                   0              0         0   
Marley Marl                                   0              0         0   
4th Disciple                                  0              0         0   
RZA                                           0              0         0   
Method Man                                    0              0         0   

              Change the Beat (Female Version)  All Night Long (Waterbed)  \
The 45 King                                  0                          0   
Marley Marl                                  0                          0   
4th Disciple                                 0                          0   
RZA                                          0                          0   
Method Man                                   0                          0   

              Fugue for Tinhorns  Funk You Up  Love the Feeling  ...  \
The 45 King                    0            0                 0  ...   
Marley Marl                    0            0                 0  ...   
4th Disciple                   0            0                 0  ...   
RZA                            0            0                 0  ...   
Method Man                     0            0                 0  ...   

              When the Morning Comes  Ain't No Half Steppin'  2 Da Good Tymz  \
The 45 King                        0                       0               0   
Marley Marl                        0                       0               0   
4th Disciple                       0                       0               0   
RZA                                0                       0               0   
Method Man                         0                       0               0   

              Kool It (Here Comes the Fuzz)  Bloodstone  Trilogy of Terror  \
The 45 King                               0           0                  0   
Marley Marl                               0           0                  0   
4th Disciple                              0           0                  0   
RZA                                       0           0                  0   
Method Man                                0           0                  0   

              How About Some Hardcore  L Dopa  Do It Your Way  Legless  
The 45 King                         0       0               0        0  
Marley Marl                         0       0               0        0  
4th Disciple                        0       0               0        0  
RZA                                 0       0               0        0  
Method Man                          0       0               0        0  

[5 rows x 7789 columns]

In [68]:
def make_util_matrix(x):
#     index = x.
    song = x.name
    new_df = df[[df['new_song_producer'] == producer] & df['sample_song_name'] == song]
    if len(new_df) > 0:
        return 1
    else:
        return 0 
    return song
    
utility_matrix.applymap(lambda x: make_util_matrix(x))

AttributeError: 'dict' object has no attribute 'applymap'

In [23]:
utility_matrix.head()

It's a New Day  We Want to Parrty, Parrty, Parrty  \
The 45 King                0                                  0   
Marley Marl                0                                  0   
4th Disciple               0                                  0   
RZA                        0                                  0   
Method Man                 0                                  0   

              Get Up, Get Into It, Get Involved  Be Black Baby  Grab Bag  \
The 45 King                                   0              0         0   
Marley Marl                                   0              0         0   
4th Disciple                                  0              0         0   
RZA                                           0              0         0   
Method Man                                    0              0         0   

              Change the Beat (Female Version)  All Night Long (Waterbed)  \
The 45 King                                  0                          0   
Marley Marl                                  0                          0   
4th Disciple                                 0                          0   
RZA                                          0                          0   
Method Man                                   0                          0   

              Fugue for Tinhorns  Funk You Up  Love the Feeling  ...  \
The 45 King                    0            0                 0  ...   
Marley Marl                    0            0                 0  ...   
4th Disciple                   0            0                 0  ...   
RZA                            0            0                 0  ...   
Method Man                     0            0                 0  ...   

              When the Morning Comes  Ain't No Half Steppin'  2 Da Good Tymz  \
The 45 King                        0                       0               0   
Marley Marl                        0                       0               0   
4th Disciple                       0                       0               0   
RZA                                0                       0               0   
Method Man                         0                       0               0   

              Kool It (Here Comes the Fuzz)  Bloodstone  Trilogy of Terror  \
The 45 King                               0           0                  0   
Marley Marl                               0           0                  0   
4th Disciple                              0           0                  0   
RZA                                       0           0                  0   
Method Man                                0           0                  0   

              How About Some Hardcore  L Dopa  Do It Your Way  Legless  
The 45 King                         0       0               0        0  
Marley Marl                         0       0               0        0  
4th Disciple                        0       0               0        0  
RZA                                 0       0               0        0  
Method Man                          0       0               0        0  

[5 rows x 7789 columns]

In [10]:
#getting to utility matrix
df.groupby(['new_song_producer', 'sampled_song_name']).head(1)
            

URL  \
0      https://www.whosampled.com/sample/126703/Big-D...   
1      https://www.whosampled.com/sample/126703/Big-D...   
3      https://www.whosampled.com/sample/126696/Big-D...   
4      https://www.whosampled.com/sample/126696/Big-D...   
6      https://www.whosampled.com/sample/3499/Big-Dad...   
7      https://www.whosampled.com/sample/4475/Big-Dad...   
8      https://www.whosampled.com/sample/73718/Big-Da...   
9      https://www.whosampled.com/sample/15527/Big-Da...   
10     https://www.whosampled.com/sample/118367/Big-D...   
11     https://www.whosampled.com/sample/592554/Big-S...   
12     https://www.whosampled.com/sample/494012/Champ...   
13     https://www.whosampled.com/sample/132635/Chill...   
14     https://www.whosampled.com/sample/13269/Chill-...   
15     https://www.whosampled.com/sample/11442/Chill-...   
16     https://www.whosampled.com/sample/286558/Chill...   
17     https://www.whosampled.com/sample/224553/12-O%...   
18     https://www.whosampled.com/sample/398325/12-O%...   
19     https://www.whosampled.com/sample/584567/Bugsy...   
20     https://www.whosampled.com/sample/316668/Cappa...   
21     https://www.whosampled.com/sample/100197/Cilva...   
22     https://www.whosampled.com/sample/266928/Cilva...   
23     https://www.whosampled.com/sample/88157/Cilvar...   
24     https://www.whosampled.com/sample/142791/Dom-P...   
25     https://www.whosampled.com/sample/165743/Dom-P...   
26     https://www.whosampled.com/sample/428446/Dom-P...   
27     https://www.whosampled.com/sample/9830/GZA-Kil...   
28     https://www.whosampled.com/sample/161622/GZA-K...   
29     https://www.whosampled.com/sample/272205/Grave...   
30     https://www.whosampled.com/sample/272205/Grave...   
31     https://www.whosampled.com/sample/273429/Grave...   
...                                                  ...   
16666  https://www.whosampled.com/sample/395373/Joell...   
16668  https://www.whosampled.com/sample/168729/Juras...   
16669  https://www.whosampled.com/sample/247306/Kirko...   
16671  https://www.whosampled.com/sample/92539/Killa-...   
16672  https://www.whosampled.com/sample/94286/Pitch-...   
16673  https://www.whosampled.com/sample/137355/Killa...   
16674  https://www.whosampled.com/sample/15875/Rakim-...   
16676  https://www.whosampled.com/sample/118208/Kool-...   
16677  https://www.whosampled.com/sample/118208/Kool-...   
16680  https://www.whosampled.com/sample/374569/Kool-...   
16681  https://www.whosampled.com/sample/374569/Kool-...   
16682  https://www.whosampled.com/sample/374569/Kool-...   
16684  https://www.whosampled.com/sample/10115/Kool-G...   
16685  https://www.whosampled.com/sample/10115/Kool-G...   
16686  https://www.whosampled.com/sample/10115/Kool-G...   
16687  https://www.whosampled.com/sample/207140/Kirko...   
16688  https://www.whosampled.com/sample/124144/Rakim...   
16689  https://www.whosampled.com/sample/24480/Kool-G...   
16690  https://www.whosampled.com/sample/24480/Kool-G...   
16691  https://www.whosampled.com/sample/24480/Kool-G...   
16692  https://www.whosampled.com/sample/23794/Rakim-...   
16693  https://www.whosampled.com/sample/86653/Kool-G...   
16694  https://www.whosampled.com/sample/86653/Kool-G...   
16695  https://www.whosampled.com/sample/86653/Kool-G...   
16696  https://www.whosampled.com/sample/21346/Rakim-...   
16698  https://www.whosampled.com/sample/48419/Kool-G...   
16699  https://www.whosampled.com/sample/18807/Kool-G...   
16700  https://www.whosampled.com/sample/183642/Kool-...   
16701  https://www.whosampled.com/sample/84049/Rakim-...   
16703  https://www.whosampled.com/sample/55090/Ras-Ka...   

                            contributor_points  \
0            Contributed by\nDukeOne (44,467 )   
1            Contributed by\nDukeOne (44,467 )   
3            Contributed by\nDukeOne (44,467 )   
4            Contributed by\nDukeOne (44,467 )   
6             Contributed by\nasymmetry (373 )   
7               Contributed by\n

In [8]:
df.head()

URL  \
0  https://www.whosampled.com/sample/126703/Big-D...   
1  https://www.whosampled.com/sample/126703/Big-D...   
3  https://www.whosampled.com/sample/126696/Big-D...   
4  https://www.whosampled.com/sample/126696/Big-D...   
6  https://www.whosampled.com/sample/3499/Big-Dad...   

                  contributor_points                  elements_sampled  \
0  Contributed by\nDukeOne (44,467 )      Direct Sample of Hook / Riff   
1  Contributed by\nDukeOne (44,467 )      Direct Sample of Hook / Riff   
3  Contributed by\nDukeOne (44,467 )  Direct Sample of Vocals / Lyrics   
4  Contributed by\nDukeOne (44,467 )  Direct Sample of Vocals / Lyrics   
6   Contributed by\nasymmetry (373 )      Direct Sample of Hook / Riff   

  name_of_contributor             new_song_album new_song_artist  \
0             DukeOne  12 Inch U.K. Release Only  Big Daddy Kane   
1             DukeOne  12 Inch U.K. Release Only  Big Daddy Kane   
3             DukeOne  12 Inch U.K. Release Only  Big Daddy Kane   
4             DukeOne  12 Inch U.K. Release Only  Big Daddy Kane   
6           asymmetry         Long Live the Kane  Big Daddy Kane   

               new_song_name new_song_producer  new_song_year  \
0  Set It Off (Extended Mix)       The 45 King           1988   
1  Set It Off (Extended Mix)       Marley Marl           1988   
3  Set It Off (Extended Mix)       The 45 King           1988   
4  Set It Off (Extended Mix)       Marley Marl           1988   
6                 Set It Off       The 45 King           1988   

  presence_of_"and throughout"_in_description sampled_artist  \
0     Sample appears at 1:24 (and throughout)    James Brown   
1     Sample appears at 1:24 (and throughout)    James Brown   
3                      Sample appears at 0:00    Lyn Collins   
4                      Sample appears at 0:00    Lyn Collins   
6     Sample appears at 0:07 (and throughout)    James Brown   

                   sampled_song_album                  sampled_song_name  \
0  It's a New Day - Let a Man Come In                     It's a New Day   
1  It's a New Day - Let a Man Come In                     It's a New Day   
3                         None Listed  We Want to Parrty, Parrty, Parrty   
4                         None Listed  We Want to Parrty, Parrty, Parrty   
6                         None Listed  Get Up, Get Into It, Get Involved   

  sampled_song_producer sampled_song_year  \
0           James Brown              1970   
1           James Brown              1970   
3           James Brown              1973   
4           James Brown              1973   
6           James Brown              1970   

  time_in_sampled_song_where_sample_appears  
0                                      0:12  
1                                      0:12  
3                                      0:00  
4                                      0:00  
6                                      3:33